### ISSUE TITLE
Clspv crashes copying structs (using llvm.memcpy) with pointer arithmetic and address space inference

### ISSUE THREAD

**Issue by BukeBeyond** [google/clspv#1173]
This may be more of a diagnostic issue.   Defaulting to global address space when inference fails, may be a valid solution.

```
struct S
{
    int i1, i2, i3; // Fails
    // int i1, i2; // OK if optimized at external compile, S is small enough not using llvm.memcpy
};

kernel void Kernel(global void* Memory)
{
    long a1 = (long) Memory;
    long a2 = a1 + sizeof(struct S); // Fails
    // long a2 = a1 + 0; // OK, optimized away, no change copy

    * (struct S*) a1 = * (struct S*) a2; // Fails
    // * (struct S*) a1 = * (global struct S*) a2; // Fails
    // * (global struct S*) a1 = * (struct S*) a2; // OK, a2 address space inferred.  Fails compiled externally.
    // * (global struct S*) a1 = * (global struct S*) a2; // OK, no need for inference
}
```

https://godbolt.org/z/KPzP7rqca

When the compilation is external, Clspv gives error:
```
immarg attribute only applies to intrinsics
ptr @_Z21llvm.memcpy.p4.p4.i64PU3AS1memcpy.p4.p4.i64
```

**Comment by rjodinchr** [comment]
This one is quite complicated

TL;DR: https://godbolt.org/z/zj9zedzf8
- `Memory` is in the global address space, thus the cast to `(struct S*)` should be `(global Struct *)`
- Because `Memory` is just used to be casted to `long`, we need to compile with `-cl-kernel-arg-info` to help clspv figure out `Memory` type. Note that clvk compiles by default with this option.
- As you are doing pointer arithmetic, you definitively need to use physical addressing (compile with `-physical-storage-buffers -arch=spir64`). We have a `LogicalPointerToIntPass` but it does not support this case. It should be able to be implemented, but that would require some work.

With all that, this example is compiled correctly.

**Comment by BukeBeyond** [comment]
Thank you for looking into this Mr. Jodin.

The interesting thing is, in CLC++ mode, Clspv infers address spaces when memcpy is not involved, as you can see below.  Clspv just converts remaining generic address spaces to global in the LowerAddrSpaceCastPass for simple loads and stores.  However, there seems to be some difficulty doing this for memcpy.  

When looking at Clang's IR from CLC++ you will see it extensively emits the generic addrspace(4), even when No function calls or generics are used!  Because the programmer of has no control of these, Clspv handling the generic spaces with memcpy is probably essential for stability.  Here are some additional observations at https://github.com/google/clspv/issues/1179   


Here the (int*) inference is working.  Also when using external Clang compilation with optimizations, S* inference works too, because it has been reduced to a single int i1.  

```
struct S { int i1; };

kernel void Kernel(global void* Memory)
{
    long a = (long) Memory;

    * (int*) a = 0; // OK, address space inferred
    // * (S*) a = {}; // Fails, address space not inferred. OK if externally optimized
    // * (global S*) a = {}; // OK, address space explicit
}

```
https://godbolt.org/z/qj4WhbMG1

Isn't it nice to have the modern syntax! :-)
